### PR TITLE
SONAR-8004 WS settings/set with a proper WS client

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java [hunk 1]
```diff
@@ -60,16 +68,79 @@ public SELF setMediaType(String s) {
...
+
+      keyValues.putAll(key, singletonList(value));
+      return this;
+    }
+
+    private DefaultParameters setValues(String key, Collection<String> values) {
+      checkArgument(!isNullOrEmpty(key));
+      checkArgument(values != null && !values.isEmpty());
+
+      this.keyValues.putAll(key, values.stream().map(Object::toString).collect(Collectors.toList()));
```
Comment (SonarTech) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java]: MAJOR NullPointerException might be thrown as 'values' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java [hunk 2]
```diff
@@ -305,27 +312,26 @@ private static SetRequest toWsRequest(Request request) {
...
+    Gson gson = GsonHelper.create();
     try {
-      return (Map<String, String>) GsonHelper.create().fromJson(json, type);
-    } catch (JsonParseException e) {
-      throw new BadRequestException(String.format("Invalid JSON '%s' for setting '%s'", json, key));
+      return (Map<String, String>) gson.fromJson(json, type);
+    } catch (JsonSyntaxException e) {
+      try {
+        gson.fromJson(json, Object.class);
+      } catch (JsonParseException invalidJson) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java]: MAJOR Either log or rethrow this exception. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/WsRequest.java [hunk 3]
```diff
@@ -32,8 +32,17 @@
...
   String getMediaType();
 
+  /**
+   *
+   * In case of multi value parameters, returns the first value
+   *
+   * @deprecated since 6.1. Use {@link #getParameters()} instead
+   */
+  @Deprecated
   Map<String, String> getParams();
```
Comment (SonarTech) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsRequest.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: it/it-tests/src/test/java/it/Category1Suite.java [hunk 4]
```diff
@@ -111,6 +111,7 @@
   @ClassRule
   public static final Orchestrator ORCHESTRATOR = Orchestrator.builderEnv()
     .setServerProperty("sonar.notifications.delay", "1")
+    .setServerProperty("sonar.web.javaAdditionalOpts", "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=8000")
```
Comment (julienlancelot) [inline it/it-tests/src/test/java/it/Category1Suite.java]: oups

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java [hunk 5]
```diff
@@ -103,7 +102,7 @@ public Builder setValues(List<String> values) {
       return this;
     }
 
-    public Builder setFieldValues(List<Map<String, String>> fieldValues) {
+    public Builder setFieldValues(List<String> fieldValues) {
```
Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: I'm wondering if we should not provide instead a ``setFieldValues(List<Map<String,String>> fieldValues)`` to hide the JSON...
But in fact, I think the goal is to map as close as possible what is done when using HTTP so I think you can forget.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java [hunk 6]
```diff
@@ -319,10 +321,17 @@ private static SetRequest toWsRequest(Request request) {
...
     Type type = new TypeToken<Map<String, String>>() {
     }.getType();
+    Gson gson = GsonHelper.create();
     try {
-      return (Map<String, String>) GsonHelper.create().fromJson(json, type);
-    } catch (JsonParseException e) {
-      throw new BadRequestException(String.format("Invalid JSON '%s' for setting '%s'", json, key));
+      return (Map<String, String>) gson.fromJson(json, type);
+    } catch (JsonSyntaxException e) {
+      try {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java]: Is this try/catch really useful ? Why displaying a different error if the JSON is not a map ?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java [hunk 7]
```diff
@@ -60,16 +68,79 @@ public SELF setMediaType(String s) {
     return (SELF) this;
   }
 
+  /**
+   * To set a multi value parameters, provide a Collection with the values
+   */
   public SELF setParam(String key, @Nullable Object value) {
```
Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java]: Why are we using Object ? Why not creating many methods with different classes ?

Comment (julienlancelot) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/BaseRequest.java]: Please move list management in another method:
```
  public SELF setParam(String key, List<Object> value) {
    Collection<Object> values = value;
    if (values.isEmpty()) {
      return (SELF) this;
    }
    parameters.setValues(key, values.stream().map(Object::toString).collect(Collectors.toList()));
    return null;
  }
```

**Comment by julienlancelot** [comment]
Apply feedbacks then LGTM

File: it/it-tests/src/test/java/it/Category1Suite.java (hunk 4 above)
Comment (teryk) [inline it/it-tests/src/test/java/it/Category1Suite.java]: oups

File: sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java (hunk 5 above)
Comment (teryk) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/setting/SetRequest.java]: as discussed offline, I've tried it but it brought dependencies in sonar-ws we don't want

**Pull request merged**